trilinos/Trilinos
# Shylu: allow fastilu to work with block CRS matrices

@trilinos/shylu 
@trilinos/tpetra 

Also, add some BlockCRS helper functions to Tpetra

## Motivation

A customer wants to be able to use shylu::fastilu with blockCrs matrices

## Testing

Correctness:

1) Enable determinism and fine-granularity fills (FASTILU_ONE_TO_ONE_UNBLOCKED). Then run soak tests on both
   CPU and GPU machines with 2 fill levels. Specifics:
   * for N = 4, 6, 8, 9
     * for block_size of each int that evenly divides N
       * for nnz % 10, 20, 50, 100
         * Generate 10 random matrices with these properties
           * For each random matrix:
             * Run Ifpack_tif_belos with these inputs
               * Run blockcrs, crs, and original (a copy of original fastilu source) FastILU impls
                 * All class members (A, L, U) should uncompress to equivalent matrices
                 * Didn't worry about convergence details

2) Run valgrind with some of these

3) Repeat the basic 6w test a bunch of times, both CPU and GPU

Performance:

For 10000x10000, 100 nnz per row, 4x4 blocks, block dense

```
CPU:
JGF initialization of blocked took: 0.0351184
JGF initialization of unblocked took: 0.805753
JGF initialization of orig took: 0.820278
 Time Init: 4.16738
JGF compute of blocked took: 0.0162119
JGF compute of unblocked took: 0.0473241
JGF compute of orig took: 0.0492717
 Finished computing Ifpack2 preconditioner
  Time (s): 0.11435
JGF apply of blocked took: 0.00204011
JGF apply of unblocked took: 0.00214537
JGF apply of orig took: 0.00173839

GPU:
JGF initialization of blocked took: 0.0791826
JGF initialization of unblocked took: 1.95681
JGF initialization of orig took: 1.93094
 Time Init: 8.77923
JGF compute of blocked took: 0.003213
JGF compute of unblocked took: 0.00254731
JGF compute of orig took: 0.00244249
 Finished computing Ifpack2 preconditioner
  Time (s): 0.0123409
JGF apply of blocked took: 0.00140019
JGF apply of unblocked took: 0.000446767
JGF apply of orig took: 0.000458988
```
